boskos/boskos.go

```diff
@@ -0,0 +1,295 @@
+		storagePath: storage,
+		JobToCattle: make(map[string]*GCPProject),
+	}
+
+	if storage != "" {
+		buf, err := ioutil.ReadFile(storage)
+		if err == nil {
+			logrus.Infof("Current state: %v.", buf)
+			err = json.Unmarshal(buf, newRanch)
+			if err != nil {
+				return nil, err
+			}
+		} else if !os.IsNotExist(err) {
+			return nil, err
+		}
+	}
+
+	file, err := ioutil.ReadFile(config)
+	if err != nil {
+		return nil, err
+	}
+
+	var data []GCPProject
+	err = json.Unmarshal(file, &data)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, p := range data {
+		for _, exist := range newRanch.JobToCattle {
+			if p.Name == exist.Name {
+				continue
+			}
+			newRanch.MyCattle = append(newRanch.MyCattle, &p)
+		}
+	}
+	newRanch.logStatus()
+
+	return newRanch, nil
+}
+
+func (r *RanchoSantaFe) logStatus() {
+	r.cattleLock.RLock()
+	defer r.cattleLock.RUnlock()
+
+	for _, proj := range r.MyCattle {
+		projJSON, _ := json.Marshal(proj)
+		logrus.Infof("Current Projects : %v", string(projJSON))
+	}
+
+	for job, proj := range r.JobToCattle {
+		projJSON, _ := json.Marshal(proj)
+		logrus.Infof("Current Map : %v - %v", job, string(projJSON))
+	}
+}
+
+func (r *RanchoSantaFe) saveState() {
+	if r.storagePath == "" {
+		return
+	}
+
+	r.cattleLock.Lock()
+	defer r.cattleLock.Unlock()
+
+	// If fail to save data, fatal and restart the server
+	buf, err := json.Marshal(r)
+	if err != nil {
+		logrus.WithError(err).Fatal("Error marshal ranch")
+	}
+	err = ioutil.WriteFile(r.storagePath+".tmp", buf, 0644)
+	if err != nil {
+		logrus.WithError(err).Fatal("Error write file")
+	}
+	err = os.Rename(r.storagePath+".tmp", r.storagePath)
+	if err != nil {
+		logrus.WithError(err).Fatal("Error rename file")
+	}
+}
+
+func (r *RanchoSantaFe) GimmeCattle(job string, duration time.Duration) *GCPProject {
+	// Handout the first free project, for duration time.
+	r.logStatus()
+	defer r.saveState()
+
+	r.cattleLock.Lock()
+	defer r.cattleLock.Unlock()
+
+	if proj, ok := r.JobToCattle[job]; ok {
+		// renew
+		proj.Duration = duration
+		proj.Start = time.Now()
+		return proj
+	}
+
+	total := len(r.MyCattle)
+	if total > 0 {
+		proj := r.MyCattle[total-1]
+		r.MyCattle = r.MyCattle[:total-1]
+		r.JobToCattle[job] = proj
+		proj.Duration = duration
+		proj.Start = time.Now()
+		return proj
+	}
+
+	return nil
+}
+
+func (r *RanchoSantaFe) ReturnCattle(job string) error {
+	// Return after finish using.
+	r.logStatus()
+	defer r.saveState()
+
+	r.cattleLock.Lock()
+	defer r.cattleLock.Unlock()
+
+	proj, ok := r.JobToCattle[job]
+	if !ok {
+		return fmt.Errorf("%v does not hold a project", job)
+	}
+	delete(r.JobToCattle, job)
+	r.MyCattle = append(r.MyCattle, proj)
+	return nil
+}
+
+func (r *RanchoSantaFe) CheckLease() bool {
+	// Callout expired leases
+	r.logStatus()
+	defer r.saveState()
+
+	r.cattleLock.Lock()
+	defer r.cattleLock.Unlock()
+
+	didSomething := false
+	for job, proj := range r.JobToCattle {
+		if proj.Duration < (time.Now().Sub(proj.Start)) {
```

Comment:
Some jobs may need to run for many hours. If this happens I would like to detect and recover from this situation faster than many hours later.